cardano-node/src/Cardano/Node/Protocol/Cardano.hs

```diff
@@ -314,11 +318,54 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
       , checkpoints = emptyCheckpointsMap
       }
 
+    -- Check that the versions for the experimental hard forks are consistent
+    warnAboutProtocolVersionsForExperimentalHardForks npcExperimentalHardForksEnabled
+      cardanoProtocolParams shelleyGenesis
+
+    return $! SomeConsensusProtocol CardanoBlockType $
+          ProtocolInfoArgsCardano cardanoProtocolParams
+
         ----------------------------------------------------------------------
         -- WARNING When adding new entries above, be aware that if there is an
         -- intra-era fork, then the numbering is not consecutive.
         ----------------------------------------------------------------------
 
+-- | Warn about the protocol versions for experimental hard forks
+warnAboutProtocolVersionsForExperimentalHardForks :: ()
+  => Bool
+  -> ProtocolParams (CardanoBlock StandardCrypto)
+  -> ShelleyGenesis StandardCrypto
+  -> ExceptT CardanoProtocolInstantiationError IO ()
+warnAboutProtocolVersionsForExperimentalHardForks npcExperimentalHardForksEnabled
+  CardanoProtocolParams
+    { hardForkTriggers = CardanoHardForkTriggers'
+        { triggerHardForkShelley = thfShelley, triggerHardForkAllegra = thfAllegra
+        , triggerHardForkMary = thfMary, triggerHardForkAlonzo = thfAlonzo
+        , triggerHardForkBabbage = thfBabbage, triggerHardForkConway = thfConway
+        }
+    } shelleyGenesis =
+  when npcExperimentalHardForksEnabled (do
+    let shelleyProtocolVersion = sgProtocolParams shelleyGenesis ^. ppProtocolVersionL
+    checkHardForkVersion "Shelley" thfShelley shelleyProtocolVersion
+    checkHardForkVersion "Allegra" thfAllegra shelleyProtocolVersion
+    checkHardForkVersion "Mary" thfMary shelleyProtocolVersion
+    checkHardForkVersion "Alonzo" thfAlonzo shelleyProtocolVersion
+    checkHardForkVersion "Babbage" thfBabbage shelleyProtocolVersion
+    checkHardForkVersion "Conway" thfConway shelleyProtocolVersion)
+  where
+    checkHardForkVersion :: ()
+      => String
+      -> Consensus.TriggerHardFork
+      -> ProtVer
+      -> ExceptT CardanoProtocolInstantiationError IO ()
+    checkHardForkVersion era (Consensus.TriggerHardForkAtVersion hfMajor) (ProtVer actualMajorVer _) =
```

Comment:
So the problem in the slack thread was hard forking to Conway at _epoch 0_. So we're actually (at this point) not interested in  `TriggerHardForkAtVersion` but we care about `TriggerHardForkAtEpoch`. We need two things:
1. A mapping of Era -> Protocol Version
2. Functions that check for the presence of the various `npcTest${era}HardForkAtEpoch` values.
Then you can compare the defined protocol version in the Shelley genesis vs what the expected protocol version. See if you can use the `Era` class in `cardano-ledger` to avoid hard coding protocol versions.

Reply:
I am comparing the Eras with the major version. But I don't understand how the epoch fits on this. How do we link the epochs to versions?

Reply:
@Jimbo4350 I am comparing the Eras with the protocol version. I am using the numbers already hardcoded in the function, and those can be overriden, apparently. So having a mapping wouldn't work? But I don't understand how the epochs fit on this. How do we link the epochs to versions? I may be completely wrong.